File: GFG POTD/019-June-2025/020-group-balls-by-sequence.cpp
// https://www.geeksforgeeks.org/problems/group-balls-by-sequence/1




#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    // TC : O(N * log(N))
    // SC : O(N)
    bool validgroup(vector<int>& arr, int k) {
        int n = arr.size();
        
        if (k == 1) return true;
        if (n % k != 0) return false;
        
        map<int, int> freq;
        for (int num : arr) {
            freq[num]++;
        }
        
        for (auto it = freq.begin(); it != freq.end(); ++it) {
            int num = it->first;
            int count = it->second;
            if (count > 0) {
                for (int i = 1; i < k; ++i) {
                    if (freq[num + i] < count) {
                        return false;
                    }
                    freq[num + i] -= count;
                }
            }
        }
        
        return true;
    }
};

File: GFG POTD/019-June-2025/020-group-balls-by-sequence.java
// https://www.geeksforgeeks.org/problems/group-balls-by-sequence/1




import java.util.*;

class Solution {
    // TC : O(N * log(N))
    // SC : O(N)
    public boolean validgroup(int[] arr, int k) {
        int n = arr.length;
        
        if (k == 1) return true;
        if (n % k != 0) return false;
        
        Map<Integer, Integer> freq = new TreeMap<>();
        for (int num : arr) {
            freq.put(num, freq.getOrDefault(num, 0) + 1);
        }
        
        for (Map.Entry<Integer, Integer> entry : freq.entrySet()) {
            int num = entry.getKey();
            int cnt = entry.getValue();
            if (cnt > 0) {
                for (int i = 1; i < k; i ++) {
                    int newNum = num + i;
                    if (freq.getOrDefault(newNum, 0) < cnt) {
                        return false;
                    }
                    freq.put(newNum, freq.get(newNum) - cnt);
                }
            }
        }
        return true;
    }
}